### DISCUSSION TITLE
What would be the best practice to synchronize new tables or new columns in old proyects

### DISCUSSION THREAD

**Discussion by maxrodro** [Mimetis/Dotmim.Sync#488] [Q&A] [answered]
Hello everyone, two days ago I discovered the great potential of Dotmin.Sync and I have been researching the existing documentation, which would be a good practice to make new versions of Xamarin as an example and update new tables or new columns.

Should it have a general Deprovision or Provision method to execute it every time a version has a new modification and it rebuilds the synchronization tables, SP, etc?

Anyone with a comment ... thank you very much

:-)

**Comment by VagueGit** [comment]
Good question @maxrodro. We are experimenting with migrating one of our apps from Realm Cloud to DMS. Realm handles migrations automagically. I think it would be great if DMS did the same.

One major downside though is that Realm migrations are additive only. This means objects/properties cannot be renamed or deprecated. I think DMS has the potential to be smarter than that.

**Comment by Mimetis** [comment, marked as answer]
Hey @maxrodro 

Actually you have a `DeprovisionAsync()` and a `ProvisionAsync()` methods.

If you look the documentation, you will see that it's not really complicated to use them : https://dotmimsync.readthedocs.io/Provision.html#provision-deprovision

The only thing to remember when using these methods is to save the scope tables at the end, depending your needs.
(_I'm thinking about simplifying this step and encapsulate the save scope in the methods, but for now, you have to do it manually_)

You have a sample here : https://github.com/Mimetis/Dotmim.Sync/tree/master/Samples/ProvisionDeprovision 

Do you need more ?

**Reply by maxrodro** [reply]
Hey @Mimetis 

Thank you for your comments, I am going to do some tests in my environment with the indicated. 👍

**Reply by maxrodro** [reply]
Hello @Mimetis ,

Several months ago we did a project in Xamarin with Dotmim with a local SQLite database connected to Azure SQL, all the synchronization functionality is perfect, now we must create a new column in a Table called EmployeesByActivities, the column is called: Quantity, the column was created and if you did the DeprovisionAsync and ProvisionAsync on the server, all good so far.

In the client (Xamarin) it is about synchronizing and it gives the following error and it is not for connection issues, it is for the creation of this new column, a Rollback is done and everything works as before.

![image](https://user-images.githubusercontent.com/64550772/134710188-0be1c3a5-7abc-4f29-bbdb-1b049b6451e4.png)

The problem that I analyze is that by creating this column on the remote server, it causes this problem; now well because the error is not clear? it should send an error from different schemas or something that helps more to identify the error.

What I can do? The moment I create the column on the remote server and do the DeprovisionAsync and ProvisionAsync, the clients will no longer be able to synchronize until that column is created on the client and does its DeprovisionAsync and ProvisionAsync locally, all the described problem is solved.

What steps should I follow or what is the best practice to do it, should I publish a new version of Xamarin, with the creation of the column locally first, once I inform myself that everyone has that version of Xamarin, I must create the column in my remote SQL and do the DeprovisionAsync and ProvisionAsync ???

Server and Client have version 0.8.0

I appreciate any comment on the subject

**Reply by Mimetis** [reply]
can you make a simple repro ?

**Reply by maxrodro** [reply]
Hi @Mimetis 

A column was simply created on the server, the DeprovisionAsync and ProvisionAsync were made and it gave the indicated error of connection problems, the created column allows null so that the old versions of the client allow to continue working normally.

**Reply by maxrodro** [reply]
Hi @Mimetis 

Any ideas that you can tell me and help me with this problem?

**Reply by Mimetis** [reply]
I need to test it, before going further.
Maybe next week.
Do you have access to the client mobile application (be able to change the code ?. )

**Reply by maxrodro** [reply]
I apreciate it!!! @Mimetis 

I have all the client code in Xamarin, I can make any changes.

**Reply by maxrodro** [reply]
Hi @Mimetis 

Have you been able to review my case?

**Reply by Mimetis** [reply]
Well, I guess the problem here is that you don't have the column names in the `SyncSetup` object, since you are syncing all the columns (and that's perfectly fine)

As a workaround, here is my try. I'm currently working on the `Customer` table, where I've added a single column:

On the server, I assume you are making the `ALTER COLUMN` on your table, before running this:

``` cs
var syncSetup = new SyncSetup(tables);
var progress = new SynchronousProgress<ProgressArgs>(s => Console.WriteLine($"{s.Source}:\t{s.Message}"));
var orchestrator = new RemoteOrchestrator(serverProvider, new SyncOptions(), syncSetup);

await orchestrator.DeprovisionAsync(SyncProvision.StoredProcedures | SyncProvision.Triggers, progress: progress);
await orchestrator.ProvisionAsync(SyncProvision.StoredProcedures | SyncProvision.Triggers, progress: progress);
```
Ok, really simple, just deprovisioning the stuff that can be impacted by columns (`StoredProcedures` and `Triggers`)
We don't need to deprovision the tracking tables since they are not impacted by the columns (only primary keys)
We don't need to deprovision the tables themselves, for ... obvious reasons :D

On the client, it's a little bit more complicated, but not that much, I guess.

If you don't want to make a full reinitialization, I guess you can do something like this. Once again, I'm assuming you are doing the `ALTER COLUMN` on the client table, before running this:

``` cs
// Fake setup to deprovision one table only
var setup = new SyncSetup(new string[] { "Customer" });

// Creating a local orchestrator with this fake setup
var localOrchestrator = new LocalOrchestrator(clientProvider, clientOptions, setup);

// Deprovision all stored procedures and triggers for the table from the fake setup
await localOrchestrator.DeprovisionAsync(SyncProvision.StoredProcedures | SyncProvision.Triggers);

// Now, getting the full schema from server
var remoteOrchestrator = new WebClientOrchestrator(serviceUri);
var serverFullSchema = await remoteOrchestrator.GetSchemaAsync().ConfigureAwait(false);

// Provision again the local client.
// Since only Customer has been deprovisioned, the only provision will occure on this table
await agent.LocalOrchestrator.ProvisionAsync(serverFullSchema).ConfigureAwait(false);
```

Obviously if you don't want to make the `ALTER TABLE` by yourself, you can delete the local SQLITE database and make a full sync again.

Can you test it and give me feedback ?

**Reply by maxrodro** [reply]
Hi @Mimetis 
Thank you very much, I am going to do the test in our environments, I appreciate your answer.
